src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java

```diff
@@ -131,6 +136,56 @@ private Cookie createRefreshTokenCookie(User user) {
         return cookie;
     }
 
+    public boolean validateTokenSignature(String token) {
+        try {
+            JWT.require(Algorithm.HMAC256(secret)).build().verify(token);
+            return true;
+        } catch (JWTVerificationException exception) {
+            System.out.println("Token verification failed: " + exception.getMessage());
+            return false;
+        }
+    }
+
+    public boolean validateTokenExpiry(String token) {
+        try {
+            DecodedJWT jwt = JWT.decode(token);
+
+            return jwt.getExpiresAt().after(new Date());
+        } catch (JWTDecodeException exception) {
+            return false;
+        }
+    }
+
+    public User getUserFromToken(String token) {
+        String userId = JWT.decode(token).getSubject();
+
+        return userRepository.findById(Long.valueOf(userId))
+                .orElseThrow(() -> new AppRuntimeException(ErrorCode.U006,
+                        "User with this id does not exist"));
+    }
+
+    public String refreshToken(HttpServletRequest request, HttpServletResponse response,
+                               String refreshToken, String accessToken) {
+        if(!validateTokenSignature(refreshToken) && !validateTokenExpiry(refreshToken)) {
+            throw new AppRuntimeException(ErrorCode.U004,
+                    "Refresh token is not valid");
+        }
+
+        String accessTokenWithoutBearer = accessToken.replace("Bearer ", "");
+        if(!validateTokenSignature(accessTokenWithoutBearer)) {
+            throw new AppRuntimeException(ErrorCode.U004,
```

Comment:
Please create a new error code in `ErrorCodes` - for example:
```java
S003("S003", "REFRESH_TOKEN_NOT_VALID", 403); // 401, 403 or other proper HTTP code.
```
**Links**
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes